File: leetcode/334.java
//Runtime: 4 ms, faster than 11.81% of Java online submissions for Increasing Triplet Subsequence.
//        Memory Usage: 38.7 MB, less than 95.35% of Java online submissions for Increasing Triplet Subsequence.
class Solution {
    public boolean increasingTriplet(int[] nums) {
        boolean res = false;
        if(nums.length < 1) return false;
        List<Integer> memo = new ArrayList<>();
        memo.add(nums[0]);
        int cur;
        for(int i=1; i<nums.length; i++) {
            if(memo.size() >= 3) break;
            if(nums[i] > memo.get(memo.size() - 1)){
                memo.add(nums[i]);
                continue;
            }
            for(int j=0; j<memo.size(); j++) {
                cur = memo.get(j);
                if(nums[i] <= cur) {
                    memo.set(j, nums[i]);
                    break;
                }
            }
        }
        return memo.size() >= 3;
    }
}

File: leetcode/334.cpp
//Runtime: 8 ms, faster than 59.20% of C++ online submissions for Increasing Triplet Subsequence.
//Memory Usage: 9.1 MB, less than 52.94% of C++ online submissions for Increasing Triplet Subsequence.
class Solution {
public:
    bool increasingTriplet(vector<int>& nums) {
        int small = INT_MAX, big = INT_MAX;
        for(int n: nums) {
            if(n <= small)
                small = n;
            else if(n <= big)
                big = n;
            else
                return true;
        }

        return false;
    }
};